boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,829 @@
+    ror  ah, 1                                  // Low 8 bits of cylinder in CH, high 2 bits
+    ror  ah, 1                                  //  in CL shifted to bits 6 & 7
+    or   cl, ah                                 // Or with sector number
+    mov  ax, HEX(0201)
+    int  HEX(13)    // DISK - READ SECTORS INTO MEMORY
+                     // AL = number of sectors to read, CH = track, CL = sector
+                     // DH = head, DL = drive, ES:BX -> buffer to fill
+                     // Return: CF set on error, AH = status (see AH=01h), AL = number of sectors read
+
+    jc   PrintDiskError                         // If the read failed then abort
+
+    popad
+
+    inc  eax                                    // Increment Sector to Read
+
+    mov  dx, es
+    add  dx, 32                                 // Increment read buffer for next sector
+    mov  es, dx
+
+    loop ReadSectorsCHSLoop                     // Read next sector
+
+    pop es
+    ret
+
+// Displays a disk error message
+// And reboots
+PrintDiskError:
+    mov  si, offset msgDiskError        // Bad boot disk message
+    call PutChars                       // Display it
+
+    jmp  Reboot
+
+// Displays a file system error message
+// And reboots
+PrintFileSystemError:
+    mov  si, offset msgFileSystemError  // FreeLdr not found message
+    call PutChars                       // Display it
+
+Reboot:
+    mov  si, offset msgAnyKey           // Press any key message
+    call PutChars                       // Display it
+    xor  ax, ax
+    int  HEX(16)                        // Wait for a keypress
+    int  HEX(19)                        // Reboot
+
+PutChars:
+    lodsb
+    or   al, al
+    jz   short Done
+    mov  ah, HEX(0e)
+    mov  bx, 7
+    int  HEX(10)
+    jmp  short PutChars
+Done:
+    ret
+
+msgDiskError:
+    .ascii "Disk error", CR, LF, NUL
+msgFileSystemError:
+    .ascii "File system error", CR, LF, NUL
+msgAnyKey:
+    .ascii "Press any key to restart", CR, LF, NUL
+
+SectsPerMFT:
+    .word 0
+    
+MFTStartSector:
+    .long 0
+
+BiosCHSDriveSize:
+    .long 0
+
+.org 509 // Pad to 509 bytes
+BootPartition:
+    .byte 1
+
+BootFlagSignature:
+    .word HEX(0aa55)    // BootSector signature
+
+// End of Bootsector
+// Next sector starts as sector 2, since boot sector is 
+// as a file under NTFS ($Boot, which has inode number 7)
+// and takes the two first clusters of the volume (which means)
+// 8 sectors (checked on Windows and mkfs.ntfs from ntfsutils)
+
+#define ATTRIBUTE_DATA HEX(80)
+#define ATTRIBUTE_INDEX_ROOT HEX(90)
+#define ATTRIBUTE_INDEX_ALLOCATION HEX(A0)
+
+StartSearch:
+    // Compute MFT start sector
+    mov   eax, dword ptr [MftLocation]
+    mov   cl, byte ptr [SectsPerCluster]
+    movzx ecx, cx
+    mul   ecx
+    mov   dword ptr MFTStartSector, eax
+
+    // Compute size of MFT entry in sectors
+    mov   al, byte ptr [ClustersPerMftRecord]
+    movzx ax, al
+    test  al, HEX(80)
+    jnz   NegativeOffset
```

Comment:
I would put a comment here, this for example https://flatcap.org/linux-ntfs/ntfs/files/boot.html

> This can be negative, which means that the size of the MFT/Index record is smaller than a cluster. In this case the size of the MFT/Index record in bytes is equal to 2^(-1 * Clusters per MFT/Index record). So for example if Clusters per MFT Record is 0xF6 (-10 in decimal), the MFT record size is 2^(-1 * -10) = 2^10 = 1024 bytes.

Just a nit, this feels more straightforward:
```
test al, al
js NegativeOffset
```